rllib/agents/a3c/a2c.py

```diff
@@ -40,6 +58,94 @@ class A2CTrainer(A3CTrainer):
     def get_default_config(cls) -> TrainerConfigDict:
         return A2C_DEFAULT_CONFIG
 
+    @override(Trainer)
+    def setup(self, config: PartialTrainerConfigDict):
+        super().setup(config)
+
+        if self.config["_disable_execution_plan_api"] is True:
+            self._microbatches = []
+
+    @override(Trainer)
+    def training_iteration(self) -> ResultDict:
+        # W/o microbatching: Identical to Trainer's default implementation.
+        # Only difference to a default Trainer being the value function loss term
+        # and its value computations alongside each action.
+        if self.config["microbatch_size"] is None:
+            return Trainer.training_iteration(self)
+
+        # In microbatch mode, we want to compute gradients on experience
+        # microbatches, average a number of these microbatches, and then
+        # apply the averaged gradient in one SGD step. This conserves GPU
+        # memory, allowing for extremely large experience batches to be
+        # used.
+        sample_batches = []
+        num_env_steps = 0
+        num_agent_steps = 0
+        while (
+            not self._by_agent_steps and num_env_steps < self.config["microbatch_size"]
+        ) or (
+            self._by_agent_steps and num_agent_steps < self.config["microbatch_size"]
+        ):
+            new_sample_batches = synchronous_parallel_sample(self.workers)
+            sample_batches.extend(new_sample_batches)
+            num_env_steps += sum(len(s) for s in new_sample_batches)
+            num_agent_steps += sum(
+                len(s) if isinstance(s, SampleBatch) else s.agent_steps()
+                for s in new_sample_batches
+            )
+        self._counters[NUM_ENV_STEPS_SAMPLED] += num_env_steps
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += num_agent_steps
+
+        # Combine all batches at once
+        train_batch = SampleBatch.concat_samples(sample_batches)
+
+        with self._timers[COMPUTE_GRADS_TIMER]:
+            grad, info = self.workers.local_worker().compute_gradients(
+                train_batch, single_agent=True
+            )
+            self._microbatches.append((grad, train_batch.count))
```

Comment:
What happens if microbatch_size is 8 and train_batch.count comes out at 32? Can this happen in some cases?
We have a similar while logic in a couple of execution plans.
In this case it would lead to a miscalculation of num_microbatches and obviously much larger microbatches?

Reply:
Yes, good point. I'll add a config check for these calculations. ...